lib/capistrano/net_storage/utils.rb

```diff
@@ -16,8 +16,9 @@ def config
       # @param dest_dir [String, Pathname] Destination directory on remote to copy local files into
       def upload_files(files, dest_dir)
         c = config
+        hosts = ::Capistrano::Configuration.env.filter(c.servers)
         if c.upload_files_by_rsync?
-          on c.servers, in: :groups, limit: c.max_parallels do |host|
+          Parallel.each(hosts, in_threads: c.max_parallels) do |host|
             ssh = build_ssh_command(host)
```

Comment:
Could you try the following code?
Although this is not something sophisticated, we can workaround the bug.

```ruby
Capistrano::NetStorage::Utils.build_ssh_command('host')
```

Maybe we can make these method public for the workaround.